### PR TITLE
[SSP][Scheduling] SSP dialect rationale and boilerplate.

### PR DESCRIPTION
**Pull request by jopperm**
Please see [`RationaleSSP.md`](https://github.com/llvm/circt/pull/3558/files#diff-3884ff41ad84300123045b6d5a1e7ba79fa569dac881dc77b099a19390dbb211).

This is the first PR in a series with the goal of landing what I drafted in #3244; please have a look there to see how this dialect's ops and attributes will be defined and how much nicer test cases/test passes will become in the future :wink:

### CODE REVIEW

File: include/circt/Dialect/SSP/SSP.td [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+include "mlir/IR/OpBase.td"
+
+def SSPDialect : Dialect {
+  let name = "ssp";
+  let summary = "Static scheduling problem instances and solutions.";
+  let description = [{
+    A dialect to abstractly represent instances and solutions of [static
+    scheduling problems](https://circt.llvm.org/docs/Scheduling/), intended as
+    an import/export format for rapid prototyping, benchmarking, and testing of
+    problem definitions and algorithm implementations.
```
Comment (mikeurbach) [inline include/circt/Dialect/SSP/SSP.td:26]: Maybe link to rationale here?

**Review by mikeurbach — APPROVED** [review]
Boilerplate looks good to me, and the rationale is quite convincing!

**Comment by mortbopet** [comment]
Continuing from #3244, i still think it would be nice to have some form of separation between the operator definitions and the actual problem graph. Perhaps the following
```mlir

ssp.operators @SSPO {
  operator_type @MemPort [latency<1>, limit<1>]
  operator_type @Add [latency<1>]
  operator_type @Implicit [latency<0>]
}

ssp.instance "canis14_fig2" of "ModuloProblem" [II<3>] {
  %0 = operation<@SSPO::@MemPort>(@store_A [dist<1>]) [t<2>]
  %1 = operation<@SSPO::@MemPort>() [t<0>]
  %2 = operation<@SSPO::@Add>(%0, %1) [t<3>]
  operation<@SSPO::@MemPort> @store_A(%2) [t<4>]
  operation<@SSPO::@Implicit>(@store_A) [t<5>]
}
```

Additionally, i think the Q&A section should contain the gist of the discussion about SSA vs symbols for auxilary dependencies. Reading the rationale as it stands now, it states nothing about what the difference between an SSA backedge and a symbol reference "backedge" is.

**Comment by jopperm** [comment]
> Continuing from #3244, i still think it would be nice to have some form of separation between the operator definitions and the problem graph. Perhaps the following [...]

Interesting. 
- (+) The design you propose would allow to re-use operator type definitions across multiple instances in handwritten test cases.
- (-) I won't be able to export/roundtrip in this form from a given `scheduling::Problem` instance; each resulting `ssp.InstanceOp` will be accompanied by its own `ssp.OperatorsOp` because the operator types are just handles and can carry different properties in every instance.
- (-) The symbol namespaces add a lot of visual redundancy, IMHO.
 
I think of the problem instances (`scheduling::` as well as `ssp`) as self-contained objects. Taking inspiration from your [earlier comment](https://github.com/llvm/circt/pull/3244#issuecomment-1141486860), I'm now considering to introduce simple container *operations* (instead of regions, which wasn't possible due to the `SymbolTable` trait), which should look the same as two regions separated by keywords. The `InstanceOp`'s region would enforce that it contains exactly one `ssp.OperatorTypesOp` and one `ssp.DependenceGraphOp`. WDYT? (I'll draft this in #3244.)

> Additionally, i think the Q&A section should contain the gist of the discussion about SSA vs symbols for auxilary dependencies. Reading the rationale as it stands now, it states nothing about what the difference between an SSA backedge and a symbol reference "backedge" is.

Good point. My plan was to include the rationale for that in the PR that adds the actual operation definitions, to keep this one short and focussed on motivating the dialect itself. I'll extend the [paragraph](https://github.com/llvm/circt/blob/34093d2d39df81934bbf86f672849788f2caf328/docs/Dialects/SSP/RationaleSSP.md#use-of-ssa-operands-and-symbol-references-to-encode-dependences) that I have prepared for this to summarize the [discussion we had](https://github.com/llvm/circt/pull/3244#issuecomment-1142157604).

**Comment by mortbopet** [comment]
> I won't be able to export/roundtrip in this form from a given scheduling::Problem instance; each resulting ssp.InstanceOp will be accompanied by its own ssp.OperatorsOp because the operator types are just handles and can carry different properties in every instance.

Definitely a valid concern.

> I'm now considering to introduce simple container operations

This makes sense. This sounds like the same approach as taken in Calyx, where `group` operation symbols - defined within the `calyx.wires` scope - can be referenced from ops within the `control` operation.

> My plan was to include the rationale for that in the PR that adds the actual operation definitions, to keep this one short and focussed on motivating the dialect itself.

Got it, thanks! 👍

**Review by mortbopet — APPROVED** [review]

**Comment by jopperm** [comment]
Thanks Morten, also for pointing out the precedent in Calyx.

**Pull request merged**